2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+Even if we alter the `DotNetAdapter` to filter out those public APIs from the experimental feature in PowerShell,
+they will still be visible to applications that host PowerShell.
+Therefore, we will not do anything about the public APIs exposed from experimental features.
+
+### Experimental Feature Discovery
+
+A user should be able to discover the available experimental features in PowerShell,
+including the ones in the PowerShell engine as well as the ones from modules.
+
+#### PowerShell Engine Experimental Feature
+
+For experimental features in PowerShell engine,
+an internal static read-only collection `EngineExperimentalFeatures` will be used to track all available engine experimental features.
+The items in the collection are of the type `ExperimentalFeature`, whose definition looks as follows:
+
+```c#
+public class ExperimentalFeature
+{
+    // Feature name
+    public string Name { get; }
+    // Feature description
+    public string Description { get; }
+    // Feature source (e.g. PSEngine, or module name)
+    public string Source { get; }
+    // Feature is enabled or not
+    public bool IsEnabled { get; }
+}
+
+internal static readonly ReadOnlyCollection<ExperimentalFeature> EngineExperimentalFeatures;
+```
+
+When adding a new experimental feature, an instance representing the feature needs to be added to the collection for it to be discoverable to users.
+The `Source` for engine experimental features should be specified as `PSEngine`.
+
+#### Module Experimental Feature
+
+For experimental features in modules,
+metadata about an experimental feature should be kept in the module manifest.
+We use the `PrivateData` section of a module manifest to expose the experimental features from the module.
+Here is an example:
+
+```powershell
+PrivateData = @{
+    ExperimentalFeatures = @(
+        @{Name = "PSWebCmdletV2", Description = "Rewrite the web cmdlets for better performance"}
+        @{Name = "PSRestCmdletV2", Description = "Rewrite the REST API cmdlets for better performance"}
+    )
+}
+```
+
+PowerShell module components, such as `Import-Module` and module analysis,
+will be updated to incorporate this metadata to the resulted `PSModuleInfo` object.
+
+#### Get-ExperimentalFeature
+
+A new cmdlet `Get-ExperimentalFeature` will be added to return all available experimental features of a PowerShell session.
+The returned experimental features are represented by the type [`ExperimentalFeature`](#powerShell-engine-experimental-feature).
+
+The cmdlet first goes through the items in `EngineExperimentalFeatures` to find all engine experimental features.
+
+Then the cmdlet goes through all loaded modules and get the available experimental features from each module.
+For each of them, the `Name` and `Description` are from the metadata. The `Source` will be the module name.
+
+Lastly, the cmdlet can go through not-yet-loaded modules that are in the module path,
+in a similar way as `Get-Command` does,
```

Comment:
Would it make more sense to have 'Get-Command -Experimental' ?

Reply:
When an experimental feature is not enabled, the cmdlets or functions subject to this feature is ignored in the command registration phase, so it doesn't make sense to have `Get-Command -Experimental`.

Reply:
Yes, but even if it is not enabled I might be interested to learn what experimental features are possible if I opt in.  I assume Get-ExperimentalFeature will show this?

Reply:
Yes, please see the type definition of `ExperimentalFeature` at [line 357](https://github.com/PowerShell/PowerShell-RFC/pull/114/files/56ba6a651b89beaf1f87e7aabd904dc68b4d4971#diff-4e7621fb8b42d3c07fd517f0a56aac9bR357). The type `ExperimentalFeature` has a property `Description` that will introduce the experimental feature.

Reply:
@PaulHigin To be clear, Get-Experimental feature enumerates the experimental features but the return type does not explicitly indicate what cmdlets, parameters, or code is impacted by the feature.  In short, we're relying on the feature author to provide an appropriate description.